### PR TITLE
feat(ecr-assets): Support cache-from and cache-to flags

### PR DESCRIPTION
**Pull request by RichiCoder1**
This adds the `--cache-from` and `--cache-to` flag options.

---

### All Submissions:
* [x]  Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md?rgh-link-date=2022-12-09T23%3A48%3A14Z)

### Adding new Construct Runtime Dependencies:
* [ ]  This PR adds new construct runtime dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md/?rgh-link-date=2022-12-09T23%3A48%3A14Z#adding-construct-runtime-dependencies)

### New Features
* [x]  Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md?rgh-link-date=2022-12-09T23%3A48%3A14Z)?

  * [x]  Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

_By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license_

### CODE REVIEW

File: packages/@aws-cdk/core/lib/assets.ts [hunk 1]
```diff
@@ -229,6 +229,18 @@ export interface DockerImageAssetSource {
...
+   * @default - no cache from args are passed
+   */
+  readonly dockerCacheFrom?: string[];
+
+  /**
+   * Cache to options to pass to the `docker build` command.
+   * @default - no cache to args are passed
+   */
+  readonly dockerCacheTo?: string;
+
```
Comment (TheRealAmazonKendra) [inline packages/@aws-cdk/core/lib/assets.ts:243]: These too

**Review by TheRealAmazonKendra — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts [hunk 2]
```diff
@@ -212,6 +212,22 @@ export interface DockerImageAssetOptions extends FingerprintOptions, FileFingerp
...
    */
   readonly outputs?: string[];
+
+  /**
+   * Cache from options to pass to the `docker build` command.
+   *
+   * @default - no cache from options are passed to the build command
+   * @see https://docs.docker.com/build/cache/backends/
+   */
+  readonly cacheFrom?: string[];
```
[the remark this answers is not in the document]
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts:222]: I was following the pattern of the other arguments as these don't have a firm schema. I think there is a general shape I could use if we want a bit more of a structure though!

[the remark this answers is not in the document]
Comment (RichiCoder1) [reply packages/@aws-cdk/aws-ecr-assets/lib/image-asset.ts:246]: @TheRealAmazonKendra I updated the PR to have a more strongishly-type cache option. Lmk how it looks!

**Comment by RichiCoder1** [comment]
Updated so that options go from:

```typescript
cacheFrom: ['type=registry,ref=<ref>']
```

to

```typescript
cacheFrom: [{ type: 'registry', params: { ref: '<ref>' }}]
```

I wasn't sure if CDK/JSII supported something like:

```typescript
interface DockerCacheOption {
  readonly type: string;
  readonly [key: string]: string;
}
```

which arguably would be more concise and still correct.

**Comment by RichiCoder1** [comment]
A potential extra evolution of this could be formalizing docker cache option as a class so you could potentially have `DockerCache.fromRegistry("myimage:main")` and `DockerCache.toInline()`, but I'm honestly not sure how that would flow through the asset system (or how symmetrical the asset information needs to be)

**Review by MrArnoldPalmer — DISMISSED** [review]
Additionally I'd like to see a unit test that asserts against the asset manifest that the build command will be formatted as expected, IE that --cache-from/to are passed correctly to the container build command.

An example of something similar https://github.com/aws/aws-cdk/blob/main/packages/%40aws-cdk/aws-ecr-assets/test/tarball-asset.test.ts#L40

Otherwise I think this looks good!

File: packages/@aws-cdk/aws-ecr-assets/test/build-image-cache.test.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    // WHEN
+    const asm = app.synth();
+
+    // THEN
+    const manifestArtifact = getAssetManifest(asm);
+    const manifest = readAssetManifest(manifestArtifact);
+
+    expect(Object.keys(manifest.dockerImages ?? {}).length).toBe(1);
+    expect(manifest.dockerImages?.[asset.assetHash]?.source.cacheFrom?.length).toBe(1);
+    expect(manifest.dockerImages?.[asset.assetHash]?.source.cacheFrom?.[0]).toStrictEqual({
```
Comment (RichiCoder1) [inline packages/@aws-cdk/aws-ecr-assets/test/build-image-cache.test.ts:27]: @MrArnoldPalmer followed the pattern in tar assets, lmk no if test naming and tests lgty!

**Review by MrArnoldPalmer — APPROVED** [review]
ahh okay I misunderstood a bit what the asset manifest contained, and thought we would be able to assert against the raw CLI command IE `docker build --cache-from mycachething .`, but this is still good coverage so TY for implementing it!

**Comment by RichiCoder1** [comment]
> ahh okay I misunderstood a bit what the asset manifest contained, and thought we would be able to assert against the raw CLI command IE `docker build --cache-from mycachething .`, but this is still good coverage so TY for implementing it!

If there's a way to accomplish that, I'd love too but I didn't see an easy way 😅

**Comment by RichiCoder1** [comment]
Found https://github.com/RichiCoder1/aws-cdk/blob/feat/ecr-assets-cache-flags/packages/cdk-assets/test/docker-images.test.ts#L390, gonna see if I can easily add a test there for the other end

**Comment by RichiCoder1** [comment]
Created #24524 as the followup PR to fix the issue I found

**Pull request merged**